logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageSender.kt

```diff
@@ -229,19 +233,26 @@ internal class MessageSenderImpl internal constructor(
             .flatMap { recipients ->
                 sessionEstablisher
                     .prepareRecipientsForNewOutgoingMessage(recipients)
-                    .map { recipients }
-                // TODO(federation) filter clients with failed to get prekeys and add persist in db
-            }.fold({
-                // TODO(federation) if (it is NetworkFailure.FederatedBackendError)
-                // TODO(federation) handle federated failure to filter clients and add to QualifiedMessageOption.IgnoreSome
-                Either.Left(it)
-            }, { recipients ->
+                    .flatMap { handleUsersWithNoClientsToDeliver(conversationId, message.id, it) }
+                    .map { recipients to it }
+            }.flatMap { (recipients, usersWithoutSessions) ->
                 messageEnvelopeCreator
                     .createOutgoingEnvelope(recipients, message)
-                    .flatMap { envelope ->
-                        trySendingProteusEnvelope(envelope, message, messageTarget)
+                    .flatMap { envelope: MessageEnvelope ->
+                        trySendingProteusEnvelope(envelope, message, messageTarget, usersWithoutSessions.users)
                     }
-            })
+            }
+    }
+
+    private suspend fun handleUsersWithNoClientsToDeliver(
+        conversationId: ConversationId,
+        messageId: String,
+        usersWithoutSessions: UsersWithoutSessions
+    ): Either<CoreFailure, UsersWithoutSessions> = if (usersWithoutSessions.hasMissingSessions()) {
+        messageRepository.persistNoClientsToDeliverFailure(conversationId, messageId, usersWithoutSessions.users)
```

Comment:
What happens if we do this for a message which isn't persisted? Calling messages for example aren't persisted into the database.

The foreign key is not found and the operation is ignored?

Reply:
Good point ! I think given that this list serves the purpose of being shown on the UI for regular messages (In which case we can assume it will be persisted) we can ignore it for the rest.